src/main/scala/mesosphere/marathon/core/volume/providers/DVDI.scala

```diff
@@ -0,0 +1,163 @@
+package mesosphere.marathon.core.volume.providers
+
+import com.wix.accord._
+import com.wix.accord.dsl._
+import com.wix.accord.Validator
+import mesosphere.marathon.core.volume._
+import mesosphere.marathon.state._
+import org.apache.mesos.Protos.{ ContainerInfo, Volume => MesosVolume, Environment }
+import scala.collection.JavaConverters._
+
+/**
+  * DVDIProvider (Docker Volume Driver Interface provider) handles persistent volumes allocated
+  * by a specific docker volume driver plugin. This works for both docker and mesos containerizers,
+  * albeit with some limitations:
+  *   - only a single volume driver per container is allowed when using the docker containerizer
+  *   - docker containerizer requires that referenced volumes be created prior to application launch
+  *   - mesos containerizer only supports volumes mounted in RW mode
+  */
+protected case object DVDIProvider extends InjectionHelper[PersistentVolume] with PersistentVolumeProvider {
+  import org.apache.mesos.Protos.Volume.Mode
+
+  override val name = Some("dvdi")
+
+  val optionDriver = name.get + "/driverName"
+  val optionIOPS = name.get + "/iops"
+  val optionType = name.get + "/volumeType"
+
+  val validOptions: Validator[Map[String, String]] = validator[Map[String, String]] { opt =>
+    opt.get(optionDriver) as "driverName option" is notEmpty
+    // TODO(jdef) stronger validation for contents of driver name
+    opt.get(optionDriver).each as "driverName option" is notEmpty
+    // TODO(jdef) validate contents of iops and volume type options
+  }
+
+  val validPersistentVolume = validator[PersistentVolume] { v =>
+    v.persistent.name is notEmpty
+    v.persistent.name.each is notEmpty
+    v.persistent.providerName is notEmpty
+    v.persistent.providerName.each is notEmpty
+    v.persistent.providerName.each is equalTo(name.get) // sanity check
+    v.persistent.options is valid(validOptions)
+  }
+
+  def nameOf(vol: PersistentVolumeInfo): Option[String] = {
+    if (vol.providerName.isDefined && vol.name.isDefined) {
+      Some(vol.providerName.get + "::" + vol.name.get)
+    }
+    else None
+  }
+
+  // group-level validation for DVDI volumes: the same volume name may only be referenced by a single
+  // task instance across the entire cluster.
+  val groupValidation: Validator[Group] = new Validator[Group] {
+    override def apply(g: Group): Result = {
+      val groupViolations = g.apps.flatMap { app =>
+        val nameCounts = volumeNameCounts(app)
+        val internalNameViolations = {
+          nameCounts.filter(_._2 > 1).map{ e =>
+            RuleViolation(app.id, s"Requested volume ${e._1} is declared more than once within app ${app.id}", None)
+          }
+        }
+        val instancesViolation: Option[RuleViolation] =
+          if (app.instances > 1) Some(RuleViolation(app.id,
```

Comment:
This must be restricted to DVDI apps. At the moment it applies to all apps, even with local docker volumes. It will always fail if there is one app running with more than one volume.